src/main/java/com/artipie/asto/s3/MultipartUpload.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 artipie.com
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.artipie.asto.s3;
+
+import com.artipie.asto.Key;
+import hu.akarnokd.rxjava2.interop.SingleInterop;
+import hu.akarnokd.rxjava2.operators.FlowableTransformers;
+import io.reactivex.Flowable;
+import io.reactivex.Observable;
+import io.reactivex.functions.Predicate;
+import java.nio.ByteBuffer;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+import org.reactivestreams.Publisher;
+import software.amazon.awssdk.core.async.AsyncRequestBody;
+import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.model.CompleteMultipartUploadRequest;
+import software.amazon.awssdk.services.s3.model.UploadPartRequest;
+import software.amazon.awssdk.services.s3.model.UploadPartResponse;
+
+/**
+ * Multipart upload of S3 object.
+ *
+ * @since 0.15
+ */
+final class MultipartUpload {
+
+    /**
+     * Approximate part size when value is uploaded as multipart.
+     */
+    private static final long PART_SIZE = 1024 * 1024;
+
+    /**
+     * S3 client.
+     */
+    private final S3AsyncClient client;
+
+    /**
+     * Bucket name.
+     */
+    private final String bucket;
+
+    /**
+     * S3 object key.
+     */
+    private final Key key;
+
+    /**
+     * ID of this upload.
+     */
+    private final String id;
+
+    /**
+     * Ctor.
+     *
+     * @param client S3 client.
+     * @param bucket Bucket name.
+     * @param key S3 object key.
+     * @param id ID of this upload.
+     * @checkstyle ParameterNumberCheck (2 line)
+     */
+    MultipartUpload(
+        final S3AsyncClient client,
+        final String bucket,
+        final Key key,
+        final String id) {
+        this.client = client;
```

Comment:
@olegmoz I think we should have a puzzle here to implement an upload data holder class. The puzzle should mention to get rid of `@checkstyle ParameterNumberCheck (2 line)`